apps/aecore/src/aec_vm_chain.erl

```diff
@@ -138,26 +138,18 @@ oracle_register(AccountKey,_Sign, QueryFee, TTL, QueryFormat, ResponseFormat,
           fee             => 0},
     {ok, Tx} = aeo_register_tx:new(Spec),
 
-    %% TODO: To register an oracle for another account than the contract
-    %%       we need a safe way to sign the register call.
-    %%       It should probably do with sign(PubKey+Nonce)
-    %%       Then we need to check that signature here.
-    %% Registering an oracle on the contract is ok.
     Result =
-        if AccountKey =:= ContractKey -> apply_transaction(Tx, State);
-           true ->
-                %% TODO: Check that Sign is correct for external accounts.
-                {error, signature_check_failed}
+        case check_account_signature(AccountKey, ContractKey, Signature) of
```

Comment:
* (A) I see `AccountKey` is the first formal parameter of the `Oracle.register` (I checked `aeso_ast_to_icode` and `aevm_ae_primops`).
  * This is published in either contract bytecode on chain, or in contract bytecode in contract create tx, or call arguments in create-or-call tx. Ok.
* (B) I see `ContractKey` is taken from the chain state that is initialized in [`aect_dispatch`](https://github.com/aeternity/epoch/blob/0118903d4d1ddbfe52bec553d4c26cd21d388ab3/apps/aecontract/src/aect_dispatch.erl#L87) from `contract` map field from [contract call tx](https://github.com/aeternity/epoch/blob/0118903d4d1ddbfe52bec553d4c26cd21d388ab3/apps/aecontract/src/aect_call_tx.erl#L230) (I did not check create tx) that is [recipient i.e. the contract whose code is being executed](https://github.com/aeternity/epoch/blob/0118903d4d1ddbfe52bec553d4c26cd21d388ab3/apps/aecontract/src/aect_call_tx.erl#L224).

In the following I focus only on contract call tx - not contract create tx. (Though similar observations may apply there too.)

I see that (B) is the address of the contract that has among its code the `CALL` instruction with arguments for calling the primop. The contract is on the chain so it is public. Ok.

I see this line does "if AccountKey equal to ContractKey then return ok otherwise check that Signature is valid signature by private key of AccountKey of binary `AccountKey++ContractKey`". Ok. So this is checking that private key of `AccountKey` is needed. Ok.

Observations: this seems to me prone to replay attacks though I may be wrong - I shall formulate some concrete examples. (I may be unable to do this before Mon.)